transmart-data/updatedb/release-17.1-HYVE-4/postgres/fix_neg_patient_nums.sql

```diff
@@ -0,0 +1,56 @@
+BEGIN;
+DO $$
+DECLARE
+min_neg_patient_num numeric;
+min_pos_patient_num numeric;
+max_pos_patient_num numeric;
+BEGIN
+	select min(patient_num) into min_neg_patient_num from i2b2demodata.patient_dimension where patient_num < 0;
+
+	IF min_neg_patient_num IS NULL THEN
+		RAISE NOTICE 'No negative patient ids found. Nothing to fix.';
+	ELSE
+		select min(patient_num) into min_pos_patient_num from i2b2demodata.patient_dimension where patient_num >= 0;
+		select max(patient_num) into max_pos_patient_num from i2b2demodata.patient_dimension where patient_num >= 0;
+		alter table i2b2demodata.relation alter constraint left_subject_id_fk deferrable initially immediate;
+		alter table i2b2demodata.relation alter constraint right_subject_id_fk deferrable initially immediate;
+		set constraints i2b2demodata.left_subject_id_fk deferred;
+		set constraints i2b2demodata.right_subject_id_fk deferred;
+		IF min_pos_patient_num IS NULL THEN
+			RAISE NOTICE 'No patients with positive identifiers were found. Using the patient id sequence instead.';
+			select nextval('i2b2demodata.SEQ_PATIENT_NUM') into min_pos_patient_num;
+			max_pos_patient_num := min_pos_patient_num;
+		END IF;
+		IF min_pos_patient_num > ABS(min_neg_patient_num) THEN
+			update i2b2demodata.patient_dimension set patient_num = min_pos_patient_num + patient_num where patient_num < 0;
```

Comment:
Impressive function! It seems that a number of tables is missing, but these are not used by us:
- `i2b2demodata.qt_patient_enc_collection`
- `i2b2demodata.archive_observation_fact`
- `deapp.de_snp_copy_number`
- `deapp.de_snp_subject_sorted_def`
- `deapp.de_snp_data_by_patient`
- `deapp.de_subject_snp_dataset`
- `deapp/de_snp_calls_by_gsm`
and some in `tm_cz`, `tm_wz`.
We could add lines here or warn the user that the migration may not cover all tables.

Reply:
Above comments lists occurences of `patient_num`. For `patient_id`, there is also the `deapp.de_subject_*_data` tables.

Reply:
and `i2b2demodata.qt_patient_sample_collection`.

Reply:
`i2b2demodata.encounter_mapping`.